spices/SPICE-0009-external-readers.adoc

```diff
@@ -0,0 +1,345 @@
+
+This proposal requires that the message passing API functionality move out of pkl-server and into pkl-core.
+The code added to pkl-core will include the new APIs and the core messaging code currently part of pkl-server (`pkl-server/src.main/kotlin/org.pkl.server/Message*.kt`).
+
+=== Message Passing API
+
+Four new message types will be added:
+
+[source,pkl]
+----
+/// Code: 0x100
+/// Type: Server Request
+class DiscoverModuleReaderRequest {
+  /// A number identifying this request
+  requestId: Int
+
+  /// The scheme of the resource to discover the spec for
+  scheme: String
+}
+
+/// Code: 0x101
+/// Type: Server Request
+class DiscoverResourceReaderRequest {
+  /// A number identifying this request
+  requestId: Int
+
+  /// The scheme of the resource to discover the spec for
+  scheme: String
+}
+
+/// Code: 0x102
+/// Type: Client Response
+class DiscoverModuleReaderResponse {
+  /// A number identifying this request
+  requestId: Int
+
+  /// Client-side module reader spec.
+  ///
+  /// Null when the external process does not implement the requested scheme.
+  /// [ClientModuleReader] is defined at https://pkl-lang.org/main/current/bindings-specification/message-passing-api.html#create-evaluator-request
+  spec: ClientModuleReader?
+}
+
+/// Code: 0x102
+/// Type: Client Response
+class DiscoverResourceReaderResponse {
+  /// A number identifying this request
+  requestId: Int
+
+  /// Client-side resource reader spec.
+  ///
+  /// Null when the external process does not implement the requested scheme.
+  /// [ClientResourceReader] is defined at https://pkl-lang.org/main/current/bindings-specification/message-passing-api.html#create-evaluator-request
+  spec: ClientResourceReader?
+}
+----
+
+`CreateEvaluatorRequest` will be expanded with additional properties:
+[source,pkl]
+----
+externalModuleReaders: Mapping<String, ExternalReader>?
+
+externalResourceReaders: Mapping<String, ExternalReader>?
+
+class ExternalReader {
+  /// May be specified as an absolute path to an executable
+  /// May also be specified as just an executable name, in which case it will be resolved according to the PATH environment variable
+  executable: String
+
+  /// Command line arguments that will be passed to the reader process
+  arguments: Listing<String>
+}
+----
+
+=== CLI
+
+New `--external-resource` and `--external-module` CLI argument will be added to configure external readers.
+The arguments can be provided multiple times to configure multiple external readers.
+The arguments are passed as `=`-delimited key-value pairs where the key is the reader's URI scheme.
+The argument values may be passed as space-separated strings where the first element becomes `executable` and any remainder becomes `arguments`.
+
+TBD: It might be best if the argument value is link:https://docs.python.org/3/library/shlex.html#shlex.split[shlex'd] instead of split to support passing arguments to the reader process that contain spaces.
```

Comment:
What does this look like?

Reply:
Let's say that an external reader is configured like so:
```
--external-resource myScheme='myExecutable arg1 "arg2 with spaces"'
```

A naive white space trim+split yields
```pkl
executable = "myExecutable"
arguments {
  "arg1"
  "\"arg2"
  "with"
  "spaces\""
}
```

While a "shlex" over the input behaves much more similarly to what a user might expect and results in
```pkl

arguments {
  "arg1"
  "arg2 with spaces"
}
```

So without the shlex operator for the CLI args it's actually not possible to express the same configs on the CLI as it is via PklProject.